pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+		kgo.FetchMaxBytes(100_000_000),
+		kgo.FetchMaxWait(5*time.Second),
+		kgo.FetchMaxPartitionBytes(50_000_000),
+		// Frequently refresh the cluster metadata. This allows us to quickly react in case some brokers are unhealthy,
+		// because the client updates the cluster info only at this frequency (and not after errors too).
+		//
+		// The side effect of frequently refreshing the cluster metadata is that if the backend returns each time a
+		// different authoritative owner for a partition, then each time cluster metadata is updated the Kafka client
+		// will client a new connection for each partition, leading to an high connections churn rate.
+		kgo.MetadataMaxAge(10*time.Second),
+		kgo.WithHooks(r.metrics.kprom),
+		kgo.WithLogger(newKafkaLogger(r.logger)),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating kafka client")
+	}
+
+	return client, nil
+}
+
+func (r *PartitionReader) fetchLastCommittedOffsetWithRetries(ctx context.Context) (offset kgo.Offset, err error) {
+	var (
+		retry = backoff.New(ctx, backoff.Config{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: 2 * time.Second,
+			MaxRetries: 10,
+		})
+	)
+
+	for retry.Ongoing() {
+		offset, err = r.fetchLastCommittedOffset(ctx)
+		if err == nil {
+			return offset, nil
+		}
+
+		level.Warn(r.logger).Log("msg", "failed to fetch last committed offset", "partition", r.partitionID, "err", err)
+		retry.Wait()
+	}
+
+	// Handle the case the context was canceled before the first attempt.
+	if err == nil {
+		err = retry.Err()
+	}
+
+	return offset, err
+}
+
+func (r *PartitionReader) fetchLastCommittedOffset(ctx context.Context) (kgo.Offset, error) {
+	cl, err := kgo.NewClient(kgo.SeedBrokers(r.kafkaAddress))
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to create admin client")
+	}
+	adm := kadm.NewClient(cl)
+	defer adm.Close()
+
+	offsets, err := adm.FetchOffsets(ctx, consumerGroup)
+	if errors.Is(err, kerr.UnknownTopicOrPartition) {
+		// In case we are booting up for the first time ever against this topic.
+		return kgo.NewOffset().AtEnd(), nil
+	}
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to fetch group offsets")
+	}
+	offset, _ := offsets.Lookup(r.kafkaTopic, r.partitionID)
+	return kgo.NewOffset().At(offset.At), nil
+}
+
+func (r *PartitionReader) commitLoop(ctx context.Context) {
+	defer r.commitLoopWg.Done()
+
+	commitTicker := time.NewTicker(r.commitInterval)
+	defer commitTicker.Stop()
+
+	toCommit := kadm.Offsets{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case fetches := <-r.commitFetches:
+			fetches.EachPartition(func(p kgo.FetchTopicPartition) {
+				if len(p.Records) == 0 {
+					return
+				}
+				r := p.Records[len(p.Records)-1]
+				toCommit.AddOffset(r.Topic, r.Partition, r.Offset+1, r.LeaderEpoch)
+			})
+		case <-commitTicker.C:
+			if len(toCommit) == 0 {
+				continue
+			}
+			committed, err := r.admClient.CommitOffsets(ctx, consumerGroup, toCommit)
```

Comment:
If this function takes > 1s (commit interval) then the ticker channel could always have a message to read and so there's no guarantee that `r.commitFetches` will be drained first. Ideally, before each call to `CommitOffset()` we want to drain `r.commitFetches`.

Reply:
That being said, I think `r.commitFetches` in a over complication. All in all you just want to keep the highest offset value per partition, and since `PartitionReader` is per-partition you just want to keep the last offset. Why not simply have an atomic int64 with the last offset to commit? We can completely remove `r.commitFetches` and just update the atomic: we want to keep the highest value between the value in the atomic, if any, and the new one to add. We can use the special value -1 to signal that the atomic is "empty" (no offset stored yet). -1 is also the value to which we reset here, right before calling `CommitOffsets()`.